### PR TITLE
feat(new sink): Add possibility to use nats jetstream in nats sink

### PR DESCRIPTION
**Pull request by whatcouldbepizza**
Add new option `jetstream` to nats sink configuration that allows using jetstream to sink into nats

### CODE REVIEW

File: src/sinks/nats/config.rs [hunk 1]
```diff
@@ -130,8 +136,56 @@ impl NatsSinkConfig {
...
+    ) -> Result<(), NatsError> {
+        match self {
+            NatsPublisher::Core(client) => {
+                client
+                    .publish(subject, payload)
+                    .await
+                    .map_err(|e| NatsError::PublishError {
+                        source: Box::new(e),
+                    })?;
+                client.flush().await.map_err(|e| NatsError::PublishError {
```
Comment (Jarema) [inline src/sinks/nats/config.rs:175]: In current version of async-nats - we no longer need to manually flush messages.
You can drop this call.

**Review by Jarema — COMMENTED** [review]
Looks good from NATS maintainer perspective.
We can also make async JetStream publishing in the future possible here.

File: src/sinks/nats/config.rs (hunk 1 above)
Comment (whatcouldbepizza) [reply src/sinks/nats/config.rs]: Thanks, done!

**Review by StephenWakely — CHANGES_REQUESTED** [review]
Awesome, thanks. Just a couple of minor nits with the code.

I'm having difficulty testing the code, possibly just down to my misunderstanding.

Running NATS in Docker:

```sh
docker run -p 4222:4222 -p 8222:8222 -p 6222:6222 --name nats-server -ti nats:latest
```

I'm pulling messages from NATS with one instance of Vector configured with:

```yml
sources:
  nats:
    type: nats
    url: "nats://localhost:4222"
    subject: nork
    connection_name: zork

sinks:
  console:
    type: console
    inputs:
      - nats
    target: stdout
    encoding:
      codec: json
```

And I'm pushing messages to NATS with another instance of Vector running this PR branch.

```yml
sources:
  in:
    type: stdin

sinks:
  nats:
    type: nats
    inputs:
      - in
    subject: nork
    url: "nats://localhost:4222"
    jetstream: true
    encoding:
      codec: json
```

I type a message into `stdin` for the NATS sink Vector instance and I get the following:

```
zork
2024-07-11T12:40:26.908904Z  WARN sink{component_kind="sink" component_id=nats component_type=nats}:request{request_id=1}: vector::sinks::util::retries: Retrying after error. error=NATS Server Error: NATS Publish Error: timed out: didn't receive ack in time internal_log_rate_limit=true
2024-07-11T12:40:32.642235Z  WARN sink{component_kind="sink" component_id=nats component_type=nats}:request{request_id=1}: vector::sinks::util::retries: Internal log [Retrying after error.] is being suppressed to avoid flooding.
2024-07-11T12:40:38.539561Z  WARN sink{component_kind="sink" component_id=nats component_type=nats}:request{request_id=1}: vector::sinks::util::retries: Internal log [Retrying after error.] has been suppressed 1 times.
2024-07-11T12:40:38.539593Z  WARN sink{component_kind="sink" component_id=nats component_type=nats}:request{request_id=1}: vector::sinks::util::retries: Retrying after error. error=NATS Server Error: NATS Publish Error: timed out: didn't receive ack in time internal_log_rate_limit=true
2024-07-11T12:40:44.042736Z  WARN sink{component_kind="sink" component_id=nats component_type=nats}:request{request_id=1}: vector::sinks::util::retries: Internal log [Retrying after error.] is being suppressed to avoid flooding.
```

The Vector instance running the NATS source receives this message repeatedly since the NATS sink keeps retrying to send:

```
{"message":"{\"host\":\"stephenwakely\",\"message\":\"zork\",\"source_type\":\"stdin\",\"timestamp\":\"2024-07-11T12:40:21.905962302Z\"}","source_type":"nats","subject":"nork","timestamp":"2024-07-11T12:40:21.908185811Z"}
{"message":"{\"host\":\"stephenwakely\",\"message\":\"zork\",\"source_type\":\"stdin\",\"timestamp\":\"2024-07-11T12:40:21.905962302Z\"}","source_type":"nats","subject":"nork","timestamp":"2024-07-11T12:40:27.641924291Z"}
{"message":"{\"host\":\"stephenwakely\",\"message\":\"zork\",\"source_type\":\"stdin\",\"timestamp\":\"2024-07-11T12:40:21.905962302Z\"}","source_type":"nats","subject":"nork","timestamp":"2024-07-11T12:40:33.539870420Z"}
{"message":"{\"host\":\"stephenwakely\",\"message\":\"zork\",\"source_type\":\"stdin\",\"timestamp\":\"2024-07-11T12:40:21.905962302Z\"}","source_type":"nats","subject":"nork","timestamp":"2024-07-11T12:40:39.042825638Z"}
{"message":"{\"host\":\"stephenwakely\",\"message\":\"zork\",\"source_type\":\"stdin\",\"timestamp\":\"2024-07-11T12:40:21.905962302Z\"}","source_type":"nats","subject":"nork","timestamp":"2024-07-11T12:40:44.557572405Z"}
```

Do you know where I am going wrong? Or is there an error in this code somewhere?

File: src/sinks/nats/config.rs [hunk 2]
```diff
@@ -76,6 +77,10 @@ pub struct NatsSinkConfig {
     #[configurable(derived)]
     #[serde(default)]
     pub(super) request: TowerRequestConfig<NatsTowerRequestConfigDefaults>,
+
+    #[configurable(derived)]
```
[the remark this answers is not in the document]
Comment (StephenWakely) [reply src/sinks/nats/config.rs:81]: Note, after adding the doc comment, you will need to run `make generate-component-docs` and then check in the resulting change to the `cue` file.

[the remark this answers is not in the document]
Comment (whatcouldbepizza) [reply src/sinks/nats/config.rs]: Changed option comment, but trying to generate docs leads to
```
/Users/dvolkov/Documents/Work/opensource/vector_fork/vector/scripts/generate-component-docs.rb:1739:in `<main>': undefined method `filter_map' for #<Hash:0x0000000122983aa0> (NoMethodError)
Did you mean?  filter
```

Is there any chance you know how to fix it?

[the remark this answers is not in the document]
Comment (jszwedko) [reply src/sinks/nats/config.rs]: I think that may indicate you are using an older version of Ruby. The script requires Ruby 3 if I remember correctly.

**Comment by whatcouldbepizza** [comment]
This may be code problem, I'll test it and come back a bit later

**Review by aliciascott — APPROVED** [review]
approved for docs

File: website/cue/reference/components/sinks/base/nats.cue [hunk 3]
```diff
@@ -383,6 +383,15 @@ base: components: sinks: nats: configuration: {
 			}
 		}
 	}
+	jetstream: {
+		description: """
```
Comment (aliciascott) [inline website/cue/reference/components/sinks/base/nats.cue:387]:
```suggestion
		description: 
```
actually these should be removed

File: website/cue/reference/components/sinks/base/nats.cue [hunk 4]
```diff
@@ -383,6 +383,15 @@ base: components: sinks: nats: configuration: {
...
+			Send messages using [Jetstream][jetstream].
+
+			[jetstream]: https://docs.nats.io/nats-concepts/jetstream
+			"""
```
Comment (aliciascott) [inline website/cue/reference/components/sinks/base/nats.cue:391]:
```suggestion
```

File: website/cue/reference/components/sinks/base/nats.cue [hunk 5]
```diff
@@ -383,6 +383,15 @@ base: components: sinks: nats: configuration: {
...
+		required: false
+		type: bool: default: false
+	}
 	request: {
 		description: """
```
Comment (aliciascott) [inline website/cue/reference/components/sinks/base/nats.cue:396]:
```suggestion
		description:
```

**Review by aliciascott — APPROVED** [review]
Hi @whatcouldbepizza I approved this but suggested to remove the `""""` from the description strings :)

File: website/cue/reference/components/sinks/base/nats.cue (hunk 3 above)
Comment (StephenWakely) [reply website/cue/reference/components/sinks/base/nats.cue:387]: This file has been generated from the doc comments added in the code, so it won't be possible to remove these quotes. They are a part of `cue` and don't make it all the way through to the final docs on the vector.dev website..

**Comment by whatcouldbepizza** [comment]
@StephenWakely unfortunately, I can't reproduce the problem, for me sinking works just fine both with and without `jetstream` option. I'll try to dive into NATS, maybe it should be configured some way to ack async publish

**Comment by whatcouldbepizza** [comment]
Hi @aliciascott 
As @StephenWakely had mentioned I did not change this file manually, it was generated by doc script. I can make suggested changes, but after next doc generating it will be overwritten, I think it will be pain to support :)

**Comment by StephenWakely** [comment]
> @StephenWakely unfortunately, I can't reproduce the problem, for me sinking works just fine both with and without `jetstream` option. I'll try to dive into NATS, maybe it should be configured some way to ack async publish
> 
> I can see that jetstream should be [configured](https://docs.nats.io/running-a-nats-service/configuration/resource_management) for nats, did you enable it?

Curious. So I hadn't enabled jetstream. So I try again, this time with it enabled:

```sh
〉docker run --rm -p 4222:4222 -p 8222:8222 -p 6222:6222 --name nats-server -ti nats:latest --jetstream                                                         nats-jetstream-option
[1] 2024/07/12 10:30:51.558624 [INF] Starting nats-server
[1] 2024/07/12 10:30:51.558695 [INF]   Version:  2.10.17
[1] 2024/07/12 10:30:51.558697 [INF]   Git:      [b91de03]
[1] 2024/07/12 10:30:51.558698 [INF]   Name:     NDUJEGWDTBUA43AXYO7FQELMWSVNUCGBOP4A7NXVD2R6Z7FZ7XIEIQSA
[1] 2024/07/12 10:30:51.558702 [INF]   Node:     Y0rIweGa
[1] 2024/07/12 10:30:51.558703 [INF]   ID:       NDUJEGWDTBUA43AXYO7FQELMWSVNUCGBOP4A7NXVD2R6Z7FZ7XIEIQSA
[1] 2024/07/12 10:30:51.558912 [INF] Starting JetStream
[1] 2024/07/12 10:30:51.559112 [INF]     _ ___ _____ ___ _____ ___ ___   _   __  __
[1] 2024/07/12 10:30:51.559116 [INF]  _ | | __|_   _/ __|_   _| _ \ __| /_\ |  \/  |
[1] 2024/07/12 10:30:51.559117 [INF] | || | _|  | | \__ \ | | |   / _| / _ \| |\/| |
[1] 2024/07/12 10:30:51.559118 [INF]  \__/|___| |_| |___/ |_| |_|_\___/_/ \_\_|  |_|
[1] 2024/07/12 10:30:51.559120 [INF]
[1] 2024/07/12 10:30:51.559121 [INF]          https://docs.nats.io/jetstream
[1] 2024/07/12 10:30:51.559123 [INF]
[1] 2024/07/12 10:30:51.559123 [INF] ---------------- JETSTREAM ----------------
[1] 2024/07/12 10:30:51.559126 [INF]   Max Memory:      23.26 GB
[1] 2024/07/12 10:30:51.559128 [INF]   Max Storage:     8.00 GB
[1] 2024/07/12 10:30:51.559129 [INF]   Store Directory: "/tmp/nats/jetstream"
[1] 2024/07/12 10:30:51.559131 [INF] -------------------------------------------
[1] 2024/07/12 10:30:51.559406 [INF] Listening for client connections on 0.0.0.0:4222
[1] 2024/07/12 10:30:51.559483 [INF] Server is ready
```

But I still get the same error. Running NATS in Docker on Ubuntu 22.04.4 LTS.

**Comment by StephenWakely** [comment]
I still get the error when running Nats outside of docker. Tried with Vector in debug and release mode.

**Comment by StephenWakely** [comment]
> Did you create a stream in NATS? Can I see its `nats stream info <stream_name>` please?

Ah ok. That was it. It now works great. Would it be possible to add this something like this to the doc comments for `jetstream`, to avoid future confusion:

```
/// If set, the `subject` must belong to an existing JetStream stream.
```

It's interesting how it managed to send the message through even though an error was returned.

**Comment by whatcouldbepizza** [comment]
I think that's the way NATS works: it something like accepts messages in `publish`, and the actual save result can be retrieved via `ack.await`

P. S. Pushed changes to field description and generated docs

**Comment by StephenWakely** [comment]
Is there a way we could verify if the stream exists in the healthcheck?

**Comment by Jarema** [comment]
This is exactly how it works - Client does not know if `Stream` exists or not when publishing. Server will return an error if ther stream is not there. 

You can check for a stream before publishing, but that it a pretty expensive call that asks the stream for it's state, config, and consesus details, so it should not be called on short interval.

**Comment by StephenWakely** [comment]
> You can check for a stream before publishing, but that it a pretty expensive call that asks the stream for it's state, config, and consesus details, so it should not be called on short interval.

It would be fine to do this in the healthcheck since that is just called once on startup. @whatcouldbepizza would you be able to make this change?

**Comment by whatcouldbepizza** [comment]
Yes, I'll do

**Comment by whatcouldbepizza** [comment]
Added nats stream checking to healthcheck if jetstream is enabled

**Comment by StephenWakely** [comment]
Unfortunately, I can't get the healthcheck to pass.

```
2024-07-16T10:10:57.377559Z ERROR vector::topology::builder: msg="Healthcheck failed." error=NATS Server Error: jetstream error: 404 (code stream not found, error code 10059) component_kind="sink" component_type="nats" component_id=nats
```

The stream does exist:

```
〉./nats stream create
? Stream Name noggle
? Subjects nork
? Storage memory
? Replication 1
? Retention Policy Limits
? Discard Policy Old
? Stream Messages Limit -1
? Per Subject Messages Limit -1
? Total Stream Size -1
? Message TTL -1
? Max Message Size -1
? Duplicate tracking time window 2m0s
? Allow message Roll-ups No
? Allow message deletion Yes
? Allow purging subjects or the entire stream Yes
Stream noggle was created

Information for Stream noggle created 2024-07-16 10:56:52

             Subjects: nork
             Replicas: 1
              Storage: Memory

Options:

            Retention: Limits
     Acknowledgements: true
       Discard Policy: Old
     Duplicate Window: 2m0s
    Allows Msg Delete: true
         Allows Purge: true
       Allows Rollups: false

Limits:

     Maximum Messages: unlimited
  Maximum Per Subject: unlimited
        Maximum Bytes: unlimited
          Maximum Age: unlimited
 Maximum Message Size: unlimited
    Maximum Consumers: unlimited


State:

             Messages: 0
                Bytes: 0 B
             FirstSeq: 0
              LastSeq: 0
     Active Consumers: 0

〉./nats stream ls
╭──────────────────────────────────────────────────────────────────────────────╮
│                                    Streams                                   │
├────────┬─────────────┬─────────────────────┬──────────┬───────┬──────────────┤
│ Name   │ Description │ Created             │ Messages │ Size  │ Last Message │
├────────┼─────────────┼─────────────────────┼──────────┼───────┼──────────────┤
│ noggle │             │ 2024-07-16 10:56:52 │ 1        │ 145 B │ 11m29s       │
╰────────┴─────────────┴─────────────────────┴──────────┴───────┴──────────────╯

```

And the sink does then work without getting any errors.

Any ideas?

**Comment by whatcouldbepizza** [comment]
What nats subject are you setting in vector sink config?

**Comment by StephenWakely** [comment]
> What nats subject are you setting in vector sink config?

```yaml
sinks:
  nats:
    type: nats
    inputs:
      - in
    subject: nork
    url: "nats://localhost:4222"
    jetstream: true
    encoding:
      codec: json
```

**Comment by whatcouldbepizza** [comment]
@Jarema I think I need NATS expert here :)
So, when we publish we provide subject only and NATS can resolve it to the relevant stream, am I right? But the `get_stream()` function takes stream name as an argument, not subject. What will be best practice here: should we iterate through all streams during `healthcheck()` and search for provided subject? Or should we add `stream` field to the config?

**Comment by whatcouldbepizza** [comment]
Since NATS subjects support wildcards, I guess it would be hard to implement the variant when we iterate through all streams and search for the subject. I'm thinking about adding new field `stream` to the config, it will be used only when `jetstream` is true, what do you think? We can even make jetstream not bool but a structure, something like
```
...
  nats:
    ...
    jetstream:
      enabled: <true/false>
      stream: <stream_name>
...
```

**Comment by StephenWakely** [comment]
> Since NATS subjects support wildcards, I guess it would be hard to implement the variant when we iterate through all streams and search for the subject. I'm thinking about adding new field `stream` to the config, it will be used only when `jetstream` is true, what do you think? We can even make jetstream not bool but a structure, something like
> 
> ```
> ...
>   nats:
>     ...
>     jetstream:
>       enabled: <true/false>
>       stream: <stream_name>
> ...
> ```

I don't think that would work since there is no way we could be sure the subject was for the given stream. Can you specify a wildcard when publishing, or is that just for subscribing to messages?

If there is no straightforward way to do this healthcheck then we can just not do it, but make sure it is made clear in the documentation.

**Comment by Jarema** [comment]
There is actually a way to filter stream list by subjects. I just need to expose it in Rust client. Will do that and make it part of a next release. That should solve the issue.

**Comment by StephenWakely** [comment]
> There is actually a way to filter stream list by subjects. I just need to expose it in Rust client. Will do that and make it part of a next release. That should solve the issue.

Oh awesome. That would be amazing! 

@whatcouldbepizza can you remove the healthcheck for now and create an issue to remind us to update it when the next version of `nats_async` is released?

**Comment by whatcouldbepizza** [comment]
Removed checking if stream exists in `healthcheck` and created the [issue](https://github.com/vectordotdev/vector/issues/20875) (for some reason I can't put any label on it, could you please decorate it with all necessary stuff like `enhancement` label and anything else if needed?)

**Comment by StephenWakely** [comment]
@whatcouldbepizza There are some clippy errors, mentioned [here](https://github.com/vectordotdev/vector/actions/runs/9976342529/job/27606967342?pr=20834). Also can you add a changelog entry to detail the change? Details can be found [here](https://github.com/vectordotdev/vector/tree/master/changelog.d).

**Comment by whatcouldbepizza** [comment]
Hi @StephenWakely! Sorry, was on a short vacation :) I've added `jetstream` option to tests and changelog file, can you run tests again please?

File: changelog.d/20834_nats_jetstream_sink.feature.md [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+  nats:
+    type: nats
+    inputs:
+      - in
+    subject: nork
+    url: "nats://localhost:4222"
+    jetstream: true
+    encoding:
+      codec: json
+```
```
Comment (StephenWakely) [inline changelog.d/20834_nats_jetstream_sink.feature.md:18]: The changelog should be fairly succinct, so we don't need the example here. We do need the authors included for external contributors.

```suggestion
authors: whatcouldbepizza
```

Comment (StephenWakely) [reply changelog.d/20834_nats_jetstream_sink.feature.md:18]: Also can you make sure it ends with a newline please.

**Comment by whatcouldbepizza** [comment]
Thanks for the comments, fixed the changelog

**Review by StephenWakely — APPROVED** [review]
Perfect! Thank you.

**Comment by StephenWakely** [comment]
Unfortunately the `nats` integration tests are now failing. It looks like not all the messages are getting through.. Can you take a look?

Provided you have Docker installed, to run the integration tests on your machine you can run `cargo vdev int test nats`.

```
thread 'sinks::nats::integration_tests::nats_tls_valid' panicked at src/sinks/nats/integration_tests.rs:58:5:
assertion `left == right` failed
  left: 8
 right: 10
stack backtrace:
   0: rust_begin_unwind
             at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/std/src/panicking.rs:652:5
   1: core::panicking::panic_fmt
             at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/core/src/panicking.rs:72:14
   2: core::panicking::assert_failed_inner
             at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/core/src/panicking.rs:409:17
   3: core::panicking::assert_failed
             at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/core/src/panicking.rs:364:5
   4: vector::sinks::nats::integration_tests::publish_and_check::{{closure}}
             at ./src/sinks/nats/integration_tests.rs:58:5
   5: vector::sinks::nats::integration_tests::nats_tls_valid::{{closure}}
             at ./src/sinks/nats/integration_tests.rs:311:37
   6: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/core/src/future/future.rs:123:9
   7: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/core/src/future/future.rs:123:9
   8: tokio::runtime::scheduler::current_thread::CoreGuard::block_on::{{closure}}::{{closure}}::{{closure}}
             at /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.38.0/src/runtime/scheduler/current_thread/mod.rs:659:57
   9: tokio::runtime::coop::with_budget
             at /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.38.0/src/runtime/coop.rs:107:5
  10: tokio::runtime::coop::budget
             at /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.38.0/src/runtime/coop.rs:73:5
  11: tokio::runtime::scheduler::current_thread::CoreGuard::block_on::{{closure}}::{{closure}}
             at /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.38.0/src/runtime/scheduler/current_thread/mod.rs:659:25
  12: tokio::runtime::scheduler::current_thread::Context::enter
             at /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.38.0/src/runtime/scheduler/current_thread/mod.rs:404:19
  13: tokio::runtime::scheduler::current_thread::CoreGuard::block_on::{{closure}}
             at /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.38.0/src/runtime/scheduler/current_thread/mod.rs:658:36
  14: tokio::runtime::scheduler::current_thread::CoreGuard::enter::{{closure}}
             at /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.38.0/src/runtime/scheduler/current_thread/mod.rs:737:68
  15: tokio::runtime::context::scoped::Scoped<T>::set
             at /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.38.0/src/runtime/context/scoped.rs:40:9
  16: tokio::runtime::context::set_scheduler::{{closure}}
             at /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.38.0/src/runtime/context.rs:180:26
  17: std::thread::local::LocalKey<T>::try_with
             at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/std/src/thread/local.rs:286:12
  18: std::thread::local::LocalKey<T>::with
             at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/std/src/thread/local.rs:262:9
  19: tokio::runtime::context::set_scheduler
             at /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.38.0/src/runtime/context.rs:180:9
  20: tokio::runtime::scheduler::current_thread::CoreGuard::enter
             at /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.38.0/src/runtime/scheduler/current_thread/mod.rs:737:27
  21: tokio::runtime::scheduler::current_thread::CoreGuard::block_on
             at /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.38.0/src/runtime/scheduler/current_thread/mod.rs:646:19
  22: tokio::runtime::scheduler::current_thread::CurrentThread::block_on::{{closure}}
             at /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.38.0/src/runtime/scheduler/current_thread/mod.rs:175:28
  23: tokio::runtime::context::runtime::enter_runtime
             at /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.38.0/src/runtime/context/runtime.rs:65:16
  24: tokio::runtime::scheduler::current_thread::CurrentThread::block_on
             at /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.38.0/src/runtime/scheduler/current_thread/mod.rs:167:9
  25: tokio::runtime::runtime::Runtime::block_on
             at /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.38.0/src/runtime/runtime.rs:347:47
  26: vector::sinks::nats::integration_tests::nats_tls_valid
             at ./src/sinks/nats/integration_tests.rs:312:5
  27: vector::sinks::nats::integration_tests::nats_tls_valid::{{closure}}
             at ./src/sinks/nats/integration_tests.rs:286:26
  28: core::ops::function::FnOnce::call_once
             at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/core/src/ops/function.rs:250:5
  29: core::ops::function::FnOnce::call_once
             at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
test sinks::nats::integration_tests::nats_tls_valid ... FAILED
```

**Comment by Jarema** [comment]
I haven't look at the test itself, but as we removed `flush` after each call (which is really not needed), it is possible to loose message if last core nats publish is immediately followed up by dropping the client.
In that case, one flush, just at the end of lifetime of the client, is recommended.

**Comment by whatcouldbepizza** [comment]
Flushing messages after driver `run()`, I think this will do the job

**Comment by StephenWakely** [comment]
Ah, apologies, I missed that in my earlier review. We do unfortunately need to flush each message after each call. This is because of our E2E acknowledgement guarantees. If someone turns that on we need to be able to guarantee to the connected source that the message has been successfully delivered.

If the message is sent without being flushed we risk passing a success back to the source, but then the flush fails and so the message isn't delivered.

Tiny bit of discussion [here](https://github.com/vectordotdev/vector/pull/18243#discussion_r1295746825).
#18272 is an issue tracking this for this sink.

Could you put the flush back after each message please?

**Comment by whatcouldbepizza** [comment]
Just put back flush after each message

**Pull request merged**